### DISCUSSION TITLE
Recovered Kicad Schematic from netlist

### DISCUSSION THREAD

**Discussion by burtonscomputers** [xjtuecho/EBAZ4205#19] [Documentation]
Sorry this is VERY specific as the first discussion started. 

I have been looking at tpecar's nl2sch project, used to generate a basic Kicad schematic from the Protel netlist.

I see in the source code he mentions :-
"    # TODO - do a simple placing algorithm that groups components based on common nets
    #        (select a central component, and the algo places everything around it)"

Can I suggest a simpler route? We could generate a list of circuit references separated into named circuit blocks.
The list could be used to group components from the same blocks together and make the job of manually relaying the schematic in a more "human-readable form", easier.
It could also make it easier for contributors to pick-up circuit blocks to work on and therefore "modularise" to work.

I had intended to jump into adding this option into the code. But having read through the source python files, I realise my python skills are WAY below those required! Hat's off to tpecar, very efficiently written!

**Comment by tpecar** [comment]
Thank you for showing interest and looking through the code!

Currently the script groups and places components based on the SchComponent (class which represents a template for a specific component).

As you've said, it shouldn't be too hard (compared to my original idea haha) to add an option where you would provide a list of references for each section, possibly as an XLS / ODS file. Each sheet in the file would represent a schematic section, and the script would group the listed components separately.

I will implement that and add it to the pull request.

In the meantime, if you have the time, you (and/or others) can provide the XLS / ODS file.

**Comment by tpecar** [comment]
It's possibly best if we do it via Google sheets, so that we can add to the same file together, and I just download the resulting file.

**Comment by tpecar** [comment]
Providing the Google sheets file

https://docs.google.com/spreadsheets/d/1fB5J5wQ95haYlHrHOwv3_8X5fgNYouUuUCoKhKYm9nM/edit?usp=sharing

It should be writable by everyone.

**Reply by burtonscomputers** [reply]
OK, making progress. Roughly 50% of the schematic parts listed. I have tried, but I fear there may be an occasional error. (Such are the risks of involving ageing humans! :-) ) After sorting it is much more difficult to cross-check. But even if a few parts are left off, it will be discovered later. I will continue on...

**Reply by tpecar** [reply]
Cool, thank you very much!

Also, the script will list if any components have been left out / marked multiple times, so it should be easy to find any issues :)

**Reply by burtonscomputers** [reply]
OK done. (at least I think so!)
We will probably be able to combine a number of the sheets later, but I think this arrangement, following the blocks on the existing pdf schematic, will ease us into the recreation of the schematics. I hope it helps.

**Reply by tpecar** [reply]
Ok, sorry it took a while, I've managed to implement the component grouping functionality.

I've used the above Google sheets file to generate the new schematic, which is now available in the pull request #18 .

Thank you very much for helping with the list! Also, you more or less nailed it ;) - there were no duplicates and only a handful missing, fixed those.

**Reply by burtonscomputers** [reply]
Really pleased you have managed this. I have downloaded your new branch, unfortunately, I was forced into a complete machine rebuild in the meantime.
Currently, the new Kicad schematic (dated 16:21 today) looks identical. (long thin page with most of the chips gathered at the top.
Also, the python program will not run to create a new schematic. Fails with this error.
![image](https://user-images.githubusercontent.com/4983112/105643037-626e7700-5e85-11eb-8a6c-1cfa823a7d43.png)

This is very likely my problem as I have rebuilt the machine from scratch in the last day or so. So am very likely to have bits missing which I had happily installed previously. Sorry! But I thought I would let you know.

**Reply by tpecar** [reply]
Hi, thanks for reviewing the files, however I believe the issues are on your side.

For the schematic, check that you're either looking at the

- [latest version of the nl2sch repo](https://github.com/tpecar/nl2sch)
- [my fork of the ebaz repository](https://github.com/tpecar/EBAZ4205/commit/8bc539a094bc44266d6b40a66140b3ab203b0722) since the pull request hasn't been included into xjtuecho's repository yet

The provided schematic looks similar as before, but is now split into sections as described by the sheet, like this

![image](https://user-images.githubusercontent.com/15571292/105646029-cc931600-5e9d-11eb-91de-bb5301963611.png)

As for the python issue, I believe it tripped on the new type hints introduced with [PEP 585](https://www.python.org/dev/peps/pep-0585/), which became available by default with Python 3.9

In earlier versions (from 3.7 onward) this functionality can be enabled by adding `from __future__ import annotations`

Try the above and report if you're still having issues.

**Reply by burtonscomputers** [reply]
Yes, totally my errors. 
1. I had not zoomed in enough to recognise the Schematic changes.
2. I only had Python 3.8 installed (Basic Ubuntu install.) So I had to get 3.91 installed manually. No that and the extra libraries are installed I have reproduced the new schematic.
Well Done!

I have made a minor edit to the "Info" sheet of Sheets Doc, deleting the "Who" entries and adding a new column for "Block Redrawn?"
I have not jumped into the schematic redraw as I want to learn more about multi-sheet schematics. But I will start to complete a few blocks in the next few days.

All the best!

**Reply by tpecar** [reply]
Ok, the script ignores the Info sheet, so it can be edited.

I've additionally sorted the entries in the Contributors section, so that the order corresponds to the sheet order (which I've also ordered by "importance/usefulness" - feel free to reorder them).

Once you lay down the framework and start with the schematic redraw, start a new thread, and we'll go from there.
